src/Databases/DatabaseOrdinary.cpp

```diff
@@ -205,17 +147,96 @@ void DatabaseOrdinary::loadStoredObjects(
+                    /// FIXME: even if we don't load the table we can still mark the uuid of it as taken.
+                    /// if (create_query->uuid != UUIDHelpers::Nil)
+                    ///     DatabaseCatalog::instance().addUUIDMapping(create_query->uuid);
+
+                    const std::string table_name = file_name.substr(0, file_name.size() - 4);
+                    LOG_DEBUG(log, "Skipping permanently detached table {}.", backQuote(table_name));
+                    return;
+                }
+
+                TableLoadingDependenciesVisitor::Data data;
+                data.default_database = metadata.default_database;
+                data.create_query = ast;
+                data.global_context = getContext();
+                TableLoadingDependenciesVisitor visitor{data};
+                visitor.visit(ast);
+                QualifiedTableName qualified_name{database_name, create_query->table};
+
+                std::lock_guard lock{metadata.mutex};
+                metadata.metadata[qualified_name] = std::make_pair(full_path.string(), std::move(ast));
+                if (data.dependencies.empty())
+                {
+                    metadata.independent_tables.emplace_back(std::move(qualified_name));
+                }
+                else
+                {
+                    for (const auto & dependency : data.dependencies)
+                    {
+                        metadata.dependencies_info[dependency].dependent_tables.push_back(qualified_name);
+                        ++metadata.dependencies_info[qualified_name].dependencies_count;
+                    }
+                }
+                metadata.total_dictionaries += create_query->is_dictionary;
+            }
+        }
+        catch (Exception & e)
+        {
+            e.addMessage("Cannot parse definition from metadata file " + full_path.string());
+            throw;
+        }
+    };
+
+    iterateMetadataFiles(local_context, process_metadata);
+
+    size_t objects_in_database = metadata.metadata.size() - prev_tables_count;
+    size_t dictionaries_in_database = metadata.total_dictionaries - prev_total_dictionaries;
+    size_t tables_in_database = objects_in_database - dictionaries_in_database;
+
+    LOG_INFO(log, "Total {} tables and {} dictionaries.", tables_in_database, dictionaries_in_database);
+}
+
+void DatabaseOrdinary::loadTableFromMetadata(ContextMutablePtr local_context, const String & file_path, const QualifiedTableName & name, const ASTPtr & ast, bool force_restore)
+{
+    assert(name.database == database_name);
+    const auto & create_query = ast->as<const ASTCreateQuery &>();
+
+    tryAttachTable(
+        local_context,
+        create_query,
+        *this,
+        name.database,
+        file_path,
+        force_restore);
 }
 
-void DatabaseOrdinary::startupTablesImpl(ThreadPool & thread_pool)
+void DatabaseOrdinary::startupTables(ThreadPool & thread_pool, bool /*force_restore*/, bool /*force_attach*/)
```

Comment:
Slightly unclear interface -- why we create pool out of this function scope, but wait it here?